### DISCUSSION TITLE
mqtt2sql config

### DISCUSSION THREAD

**Discussion by TelecomServices** [curzon01/mqtt2mysql#19] [General]
Hello,

I have installed the the scripts.
Now i have a problem, i have a error message in line 100:
mqtt-exclude-topic: None
in the mqtt2sql.py 
    'mqtt_url': 'mqtt://localhost/#',
    'mqtt-topic': devices/id_MEM6REM6N0U6RTU6NTE6MDg/#,
    'mqtt-exclude-topic': None,
    'mqtt-cafile': None,
    'mqtt-certfile': None,
    'mqtt-keyfile': None,
    'mqtt-insecure': False,
    'mqtt-keepalive': 60,
    'mqtt-connect-timeout':500,

    'sql-type': 'mysql' if MODULE_MYSQLDB_AVAIL else ('sqlite' if MODULE_SQLITE3_AVAIL else None),
    'sql-host': 'localhost',
    'sql-port': 3306,
    'sql-username': adminer,
    'sql-password': TSO4546a,
    'sql-db': mqtt,
    'sql-table': 'mqtt',
    'sql-max-connection': 50,
    'sql-connection-retry': 10,
    'sql-connection-retry-start-delay': 1,
    'sql-transaction-retry': 10

I do not know what is wrong.
I do not use a username and password in the mqtt.
No exclude topics
no certfile or keyfile

**Comment by curzon01** [comment]
You edited the program and also in a wrong way.

To call a local program copy don't touch mqtt2sql.py, use parameter and/or a conf file as [described here](https://github.com/curzon01/mqtt2sql#usage).

To start the unchanged copy of mqtt2sql.py using your own parameter either

a) using Program parameter
`./mqtt2sql.py --mqtt-url 'mqtt://localhost/#' --mqtt-topic 'devices/id_MEM6REM6N0U6RTU6NTE6MDg/#' --sql-username adminer --sql-password TSO4546a`

b) altenatively put the parameter into a conf file e.g. create a text file called `my.conf` with the content:
```
mqtt-url = 'mqtt://localhost/#'
mqtt-topic = 'devices/id_MEM6REM6N0U6RTU6NTE6MDg/#' 
sql-username = adminer
sql-password = TSO4546a
```

and start the program using

`./mqtt2sql.py -c my.conf`

To get a complete list of parameter use `./mqtt2sql.py -h`

**Reply by TelecomServices** [reply]
When i use your ./mqtt2sql.py --mqtt-url............. i get the message:
mqtt2sql.py: error: unrecognized arguments: --mqtt-url mqtt://localhost/#

**Reply by TelecomServices** [reply]
With my mqtt explorer i can connect to the mqtt broker.
And i have tested if i can send messages:
mosquitto_sub -h 127.0.0.1 -t topic

Open another terminal and type
mosquitto_pub -h 127.0.0.1 -t topic -m "Hello"

This is working so i can make a connection over ip adres 127.0.0.1 and localhost

**Comment by curzon01** [comment]
sorry, your program changes confused me, mqtt_url is an internal identifier not a valid (unrecognized) argument but reading `./mqtt2sql.py -h` would have helped:

```
  --mqtt <URL>          Specify specify user, password, hostname, port and topic at once as a URL (default 'mqtt://localhost/#').
                        The URL must be in the form:
```

so use `./mqtt2sql.py --mqtt 'mqtt://localhost/#' ...`

BTW:
>  i can make a connection over ip adres 127.0.0.1 and localhost

your `mosquitto_pub` examples shows me that you can connect to 127.0.0.1, it does not show that `localhost`(used with mqtt2sql) is resolved to `127.0.0.1` (used with mosquitto_pub)

if 127.0.0.1 works for `mosquitto_pub` use same for mqtt2sql
if you would use 'localhost' and it doesn't work with mqtt2sql, try localhost for `mosquitto_pub` too
otherwise you change two things in one step (program and address) and don't know where the mistake is.


PS: I updated the program with a small fix but that's unrelated to your question.

**Comment by TelecomServices** [comment]
Hi,
After your update is my problem solved, i have download the files again and now it is working.
Thank you.
It looks nice.
